httpx/content_streams.py

```diff
@@ -0,0 +1,283 @@
+
+        def guess_content_type(self) -> typing.Optional[str]:
+            if self.filename:
+                return (
+                    mimetypes.guess_type(self.filename)[0] or "application/octet-stream"
+                )
+            else:
+                return None
+
+        def render_headers(self) -> bytes:
+            parts = [
+                b"Content-Disposition: form-data; ",
+                format_form_param("name", self.name),
+            ]
+            if self.filename:
+                filename = format_form_param("filename", self.filename)
+                parts.extend([b"; ", filename])
+            if self.content_type is not None:
+                content_type = self.content_type.encode()
+                parts.extend([b"\r\nContent-Type: ", content_type])
+            parts.append(b"\r\n\r\n")
+            return b"".join(parts)
+
+        def render_data(self) -> bytes:
+            if isinstance(self.file, str):
+                content = self.file
+            else:
+                content = self.file.read()
+            return content.encode("utf-8") if isinstance(content, str) else content
+
+    def __init__(self, data: dict, files: dict, boundary: bytes = None) -> None:
+        body = BytesIO()
+        if boundary is None:
+            boundary = binascii.hexlify(os.urandom(16))
+
+        for field in self.iter_fields(data, files):
+            body.write(b"--%s\r\n" % boundary)
+            body.write(field.render_headers())
+            body.write(field.render_data())
+            body.write(b"\r\n")
+
+        body.write(b"--%s--\r\n" % boundary)
+
+        self.content_type = "multipart/form-data; boundary=%s" % boundary.decode(
+            "ascii"
+        )
+        self.body = body.getvalue()
+
+    def iter_fields(
+        self, data: dict, files: dict
+    ) -> typing.Iterator[typing.Union["FileField", "DataField"]]:
+        for name, value in data.items():
+            if isinstance(value, (list, dict)):
+                for item in value:
+                    yield self.DataField(name=name, value=item)
+            else:
+                yield self.DataField(name=name, value=value)
+
+        for name, value in files.items():
+            yield self.FileField(name=name, value=value)
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        content_length = str(len(self.body))
+        content_type = self.content_type
+        return {"Content-Length": content_length, "Content-Type": content_type}
+
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        yield self.body
+
+
+def encode(
+    data: RequestData = None,
+    files: RequestFiles = None,
+    json: typing.Any = None,
+    boundary: bytes = None,
+) -> ContentStream:
+    """
+    Handles encoding the given `data`, `files`, and `json`, returning
+    a `ContentStream` implementation which provides a byte iterator onto
+    the content, as well as `.is_rewindable()` and `.get_headers()` interfaces.
```

Comment:
Small inconsistency on `.is_rewindable()` here. I think slimming this down to the following is sufficient:

```
    Handles encoding the given `data`, `files`, and `json`, returning
    a `ContentStream` implementation.
```